### ISSUE TITLE
[naming-convention] camelCase allows kebab-case in property

### ISSUE THREAD

**Issue by koooge** [typescript-eslint/typescript-eslint#4534]
Hi there,

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```js
{
  "rules": {
    camelcase: 'off',
    '@typescript-eslint/naming-convention': [
      'error',
      {
        selector: 'default',
        format: ['camelCase'],
        leadingUnderscore: 'allow',
        trailingUnderscore: 'allow',
      },
      {
        selector: 'variable',
        format: ['camelCase', 'UPPER_CASE'],
        leadingUnderscore: 'allow',
        trailingUnderscore: 'allow',
      },
      {
        selector: 'typeLike',
        format: ['PascalCase'],
      },
    ],
  },
};
```

```TS
type Foo = {
  'foo-bar': string;
};

interface Bar {
  'foo-bar': string;
}
```

**Expected Result**
It is supposed to be error:

`1: 3 error Type property name 'foo-bar'  must match one of the following formats: camelCase @typescript-eslint/naming-convention`

**Actual Result**
No error.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.10.0` |
| `@typescript-eslint/parser`        | `5.10.0` |
| `TypeScript`                       | `4.5.4` |
| `ESLint`                           | `8.7.0` |
| `node`                             | `16.13.2` |

**Comment by bradzacher** [comment]
I was going to say that this was intentional, but then I noticed the examples I'd placed in the readme work via sheer dumb luck!
The examples all use `strictCamelCase`, and then have strings that start with an upper case letter - so they violate the format not because of spaces or dashes, but because of that first letter 🤦 .

The current checkers for pre-defined formats by design assume the name is a "valid identifier" (i.e. no spaces or dashes).
https://github.com/typescript-eslint/typescript-eslint/blob/27946318b0b4ad7fb2d17428b4b335e6827032d1/packages/eslint-plugin/src/rules/naming-convention-utils/format.ts#L17-L41

We should probably make each one also enforce that the string is also a valid identifier using this util:
https://github.com/typescript-eslint/typescript-eslint/blob/27946318b0b4ad7fb2d17428b4b335e6827032d1/packages/type-utils/src/requiresQuoting.ts

**Comment by lonyele** [comment]
@bradzacher  I was looking at the code and found the fix looks simple which makes me think that I didn't fully understand your approach. 
If I add this `requiresQuoting` check to this place
https://github.com/typescript-eslint/typescript-eslint/blob/5b7d8df149be6554c863cdd5b73c0b0b0a7960db/packages/eslint-plugin/src/rules/naming-convention-utils/validator.ts#L385-L390

like this. 
```ts
for (const format of formats) {
  const checker = PredefinedFormatToCheckFunction[format];
  if (!util.requiresQuoting(name) && checker(name)) {
    return true;
  }
}
```

I can pass other tests and get this issue's expected result
`'Type Property name 'foo       Bar' must match one of the following formats: camelCase'` or 
`'Type Property name 'boo-----foo' must match one of the following formats: camelCase'`

I could have guarded  using `requiresQuoting` inside of all 6 predefined formats functions but I chose this one because it looks simpler when no other places are using these predefined formats

do you think this is the viable fix?

**Comment by bradzacher** [comment]
Nope - you're right! It should be super simple to do!
Ideally we wouldn't call `util.requiresQuoting` on every format because it does some really complicated regex checks

Given that where it counts we already add a modifier to the name which specifies whether or not it requires quotes:
https://github.com/typescript-eslint/typescript-eslint/blob/877cc48590c92cba7ff04c0cf421af50de9acab5/packages/eslint-plugin/src/rules/naming-convention.ts#L114-L116
https://github.com/typescript-eslint/typescript-eslint/blob/877cc48590c92cba7ff04c0cf421af50de9acab5/packages/eslint-plugin/src/rules/naming-convention.ts#L423-L425

We know all of our pre-defined formats are valid identifier string (so don't require quoting). Which means all we should need to do is pass the `modifiers` into `validatePredefinedFormat` and then check like `if (modifiers.has(Modifiers.requiresQuotes)) { return false }`.

It should be really simple!
